lib/stdio/native.js

```diff
@@ -0,0 +1,44 @@
+import process from 'node:process';
+import {isStream as isNodeStream} from 'is-stream';
+
+// When we use multiple `stdio` values for the same streams, we pass 'pipe' to `child_process.spawn()`.
+// We then emulate the piping done by core Node.js.
+// To do so, we transform the following values:
+//  - Node.js streams are marked as `type: nodeStream`
+//  - 'inherit' becomes `process.stdin|stdout|stderr`
+//  - any file descriptor integer becomes `process.stdio[index]`
+// All of the above transformations tell Execa to perform manual piping.
+export const handleNativeStream = (stdioStream, isStdioArray) => {
+	const {type, value, index, optionName} = stdioStream;
+
+	if (!isStdioArray || type !== 'native') {
+		return stdioStream;
+	}
+
+	if (value === 'inherit') {
+		return {...stdioStream, type: 'nodeStream', value: getStandardStream(index, value, optionName)};
+	}
+
+	if (typeof value === 'number') {
+		return {...stdioStream, type: 'nodeStream', value: getStandardStream(value, value, optionName)};
+	}
+
+	if (isNodeStream(value)) {
+		return {...stdioStream, type: 'nodeStream'};
+	}
+
+	return stdioStream;
+};
+
+// Node.js does not allow to easily retrieve file descriptors beyond stdin/stdout/stderr as streams
```

Comment:
I discovered this when doing this PR. 
While Node.js has `childProcess.stdio[index]` to access any file descriptor as a stream, it does not have a way to do so on the current process, beyond file descriptors 0/1/2? 
I looked into the `fs` module, but although many methods can take a file descriptor as argument, and although `FileHandle` exists, there does not seem to be a way to open a file descriptor as a Node.js stream (or am I missing something?)

Reply:
Did you try `fs.createReadStream('', {fd: 4})`?

Reply:
That was a very good tip, I forgot this existed!

Unfortunately, there is the following problem.

The stream created by `fs.createReadStream()` uses `fs.read(fd, ...)` under the hood. However, when the file descriptor points to an interactive TTY (or any other character device with blocking reads), `fs.read()` might never resolve, which makes the process hang. Even after calling both `stream.destroy()` and `fs.close(fd)`, `fs.read()` keeps going and holding the process.

https://github.com/nodejs/node/blob/345f15e493162e27fd8b543b7bc995aa45469639/lib/internal/fs/streams.js#L261

This can be reproduced by running this in an interactive TTY. The process hangs.

```js
const stream = createReadStream('', {fd: 0});
stream.read()
setTimeout(() => {
  stream.destroy();
}, 0);
```

The Node.js documentation for [`fs.createReadStream()`](https://nodejs.org/api/fs.html#fscreatereadstreampath-options) warns about this:

> If `fd` points to a character device that only supports blocking reads (such as keyboard or sound card), read operations do not finish until data is available. This can prevent the process from exiting and the stream from closing naturally.

I have been digging into the Node.js source code for `fs.createReadStream()` to find a solution, but I am not sure there is one. :thinking: If you think of anything, please let me know!

Reply:
Maybe ChatGPT can help:


Your analysis of the issue with `fs.createReadStream()` and blocking reads on character devices like TTYs is accurate. The stream's underlying `fs.read()` call can cause the process to hang since it waits for data to be available. Unfortunately, Node.js does not provide a built-in, straightforward way to handle this scenario with its standard APIs.

However, you can approach this problem using non-blocking techniques. One potential solution is to use `net.Socket` with the file descriptor. This allows you to handle TTY or other character devices in a non-blocking manner. The `net.Socket` class can be used to wrap a file descriptor and convert it into a stream that behaves more predictably with character devices.

Here's an example of how you might implement this:

```javascript
import { Socket } from 'net';

// Wrap the file descriptor in a net.Socket
const fd = 0; // Standard input (stdin)
const socket = new Socket({ fd, readable: true, writable: false });

// Read data from the socket
socket.on('data', (data) => {
    console.log('Received:', data.toString());
});

// Handle any errors
socket.on('error', (err) => {
    console.error('Socket error:', err);
});

// Example of cleaning up
setTimeout(() => {
    socket.destroy();
}, 5000); // Destroys the socket after 5 seconds
```

In this code, we create a `Socket` object and pass the file descriptor along with readability and writability options. This socket acts as a stream and provides the typical stream events like `data`, `error`, and `end`.

Using `net.Socket` is a more advanced technique and requires careful handling, especially regarding the lifetime and cleanup of the socket. Be aware that this approach may have nuances depending on the specifics of the file descriptor and the environment in which your Node.js application is running.


Another solution to handle non-blocking reads from a file descriptor in Node.js, especially for character devices like TTYs, is to use `polling` with the `fs.read` method. The idea is to repeatedly check if there's data available on the file descriptor without blocking the event loop.

Here's an approach using `setImmediate` or `setTimeout` to periodically poll the file descriptor:

1. **Use `fs.read` in Non-blocking Mode**: Call `fs.read` in a loop, but ensure it doesn't block the event loop. You can do this by using `setImmediate` or `setTimeout`.

2. **Manage the Read Buffer**: Handle the buffer management yourself.

3. **Proper Cleanup**: Ensure to clean up and close the file descriptor when done.

Here's a basic example to illustrate this:

```javascript
import { open, read, close } from 'fs';

// Open the file descriptor for reading (e.g., fd 0 for stdin)
const fd = 0; // Using stdin as an example

// Function to perform a non-blocking read
function nonBlockingRead() {
    const buffer = Buffer.alloc(1024);
    read(fd, buffer, 0, buffer.length, null, (err, bytesRead, buffer) => {
        if (err) {
            console.error('Error reading:', err);
            return;
        }

        if (bytesRead > 0) {
            console.log('Read:', buffer.slice(0, bytesRead).toString());
        }

        // Continue polling
        setImmediate(nonBlockingRead);
    });
}

// Start the non-blocking read loop
nonBlockingRead();

// Example cleanup after some time
setTimeout(() => {
    close(fd, (err) => {
        if (err) {
            console.error('Error closing file descriptor:', err);
        } else {
            console.log('File descriptor closed successfully.');
        }
    });
}, 10000); // 10 seconds
```

In this example, `nonBlockingRead` is a function that reads from the file descriptor in a non-blocking way. It uses `setImmediate` to continuously poll the file descriptor. The `setTimeout` is used just as an example to stop the polling and close the file descriptor after a certain period.

This approach gives you more control over how you read from the file descriptor, but it requires careful management of the buffer and the read loop. Additionally, this method is somewhat low-level and may not be as convenient as using built-in stream abstractions. Keep in mind that for very high-frequency data, this approach might still introduce performance overhead due to the repeated scheduling of callbacks.

Reply:
I don't think the polling method ChatGPT is suggesting works. I do not think `fs.read()` has a polling option, and the example suggested uses the same API that is used by `fs.createReadStream()`. In other words, `fs.read()` will not return right away if there is no data available, it will block.

Now, the `Socket` suggestion would work. However, opening and managing a TCP socket seems quite heavy for the edge case we're covering here.

We could also detect whether the file descriptor points to a character device, and only throw in that case. However, it would lead to consumers to have their code fail when redirecting to an interactive terminal, but not in other cases, which will most likely cause confusion.

The case we're trying to cover is the following: `execa(..., { stdio: [..., ..., ..., ['inherit', ...])`. This requires the combination of all of:
  1. A custom file descriptor is used (beyond `stdin`/`stdout`/`stderr`)
  2. That custom file descriptor uses an array of multiple values, such as `['inherit', 'pipe']`
  3. One of those values is either `'inherit'`, or a file descriptor integer that is not 0/1/2

This is a very unlikely edge case, which might not justify adding too much complexity to the implementation. I think we might be good with just keeping it as is, and throwing a helpful error message. What do you think?

Reply:
> and throwing a helpful error message. I do not even think this might be worth documenting in the readme.md, as almost no users would run into that use case, and it might confuse them more than help.

Agreed

Reply:
This should be done now. I have also added some code comments to summarize the above.